File: leet_code/ruby/contiguous_array.rb
# @param {Integer[]} nums
# @return {Integer}
def find_max_length(nums)
	map = { 0 => -1 }
	max, count = 0, 0
	(0...nums.size).each do |i|
		count += nums[i] == 1 ? 1 : -1
		if map.key? count
			max = [max, i - map[count]].max
		else
			map[count] = i
		end
	end
	max
end

File: leet_code/go/contiguousArray.go
func findMaxLength(nums []int) int {
	hash := make(map[int]int)
	hash[0] = -1
	max, count := 0, 0
	for i := 0; i < len(nums); i++ {
		if nums[i] == 1 {
			count += 1
		} else {
			count += -1
		}
		if _, ok := hash[count]; ok {
			if max < i - hash[count] {
				max = i - hash[count]
			}
		} else {
			hash[count] = i
		}
	}
	return max
}